overlord/patch/patch6_2.go

```diff
@@ -0,0 +1,168 @@
+}
+
+type patch62Flags struct {
+	DevMode          bool `json:"devmode,omitempty"`
+	JailMode         bool `json:"jailmode,omitempty"`
+	Classic          bool `json:"classic,omitempty"`
+	TryMode          bool `json:"trymode,omitempty"`
+	Revert           bool `json:"revert,omitempty"`
+	RemoveSnapPath   bool `json:"remove-snap-path,omitempty"`
+	IgnoreValidation bool `json:"ignore-validation,omitempty"`
+	Required         bool `json:"required,omitempty"`
+	SkipConfigure    bool `json:"skip-configure,omitempty"`
+	Unaliased        bool `json:"unaliased,omitempty"`
+	Amend            bool `json:"amend,omitempty"`
+	IsAutoRefresh    bool `json:"is-auto-refresh,omitempty"`
+	NoReRefresh      bool `json:"no-rerefresh,omitempty"`
+	RequireTypeBase  bool `json:"require-base-type,omitempty"`
+}
+
+type patch62SnapState struct {
+	SnapType string             `json:"type"`
+	Sequence []*patch62SideInfo `json:"sequence"`
+	Active   bool               `json:"active,omitempty"`
+	Current  snap.Revision      `json:"current"`
+	Channel  string             `json:"channel,omitempty"`
+	patch62Flags
+	Aliases              interface{} `json:"aliases,omitempty"`
+	AutoAliasesDisabled  bool        `json:"auto-aliases-disabled,omitempty"`
+	AliasesPending       bool        `json:"aliases-pending,omitempty"`
+	UserID               int         `json:"user-id,omitempty"`
+	InstanceKey          string      `json:"instance-key,omitempty"`
+	CohortKey            string      `json:"cohort-key,omitempty"`
+	RefreshInhibitedTime *time.Time  `json:"refresh-inhibited-time,omitempty"`
+}
+
+type patch62SnapSetup struct {
+	Channel   string    `json:"channel,omitempty"`
+	UserID    int       `json:"user-id,omitempty"`
+	Base      string    `json:"base,omitempty"`
+	Type      snap.Type `json:"type,omitempty"`
+	PlugsOnly bool      `json:"plugs-only,omitempty"`
+	CohortKey string    `json:"cohort-key,omitempty"`
+	Prereq    []string  `json:"prereq,omitempty"`
+	patch62Flags
+	SnapPath     string           `json:"snap-path,omitempty"`
+	DownloadInfo interface{}      `json:"download-info,omitempty"`
+	SideInfo     *patch62SideInfo `json:"side-info,omitempty"`
+	patch62auxStoreInfo
+	InstanceKey string `json:"instance-key,omitempty"`
+}
+
+type patch62auxStoreInfo struct {
+	Media interface{} `json:"media,omitempty"`
+}
+
+// patch6_2:
+//  - ensure snapd snaps in the snapstate have TypeSnapd for backward compatibility with old snapd snap releases.
+//  - ensure snapd snaps have TypeSnapd in pending install tasks.
+func patch6_2(st *state.State) error {
+	var snaps map[string]*json.RawMessage
+	if err := st.Get("snaps", &snaps); err != nil && err != state.ErrNoState {
+		return fmt.Errorf("internal error: cannot get snaps: %s", err)
+	}
+
+	snapStates := make(map[string]*patch62SnapState)
+	var haveSnapdSnap bool
+
+	// Migrate snapstate; bail out early if we detect snapd snap.
+	// Break up the loop over snaps into two to ease testing.
+	for name, raw := range snaps {
```

Comment:
(nitpick^2) I would love something slightly simpler here. I'm not sure the added complexity with the two loops is worth it. The idea is nice because it makes testing for the "breaks" possible but constructing the test-case needs an "impossible" state.json. Plus one could argue that really all snapd-snap-id snaps should be converted to the TypeSnapd - or am I missing something?

Anyway my staw-man would be something like below and drop the TestPatch62StopsIfSnapdAlreadyPresent:
```go
func hasSnapdSnapID(snapst patch62SnapState) bool {
	for _, seq := range snapst.Sequence {
		if snap.IsSnapd(seq.SnapID) {
			return true
		}
	}
	return false
}

// patch6_2:
//  - ensure snapd snaps in the snapstate have TypeSnapd for backward compatibility with old snapd snap releases.
//  - ensure snapd snaps have TypeSnapd in pending install tasks.
func patch6_2(st *state.State) error {
	var snaps map[string]*json.RawMessage
	if err := st.Get("snaps", &snaps); err != nil && err != state.ErrNoState {
		return fmt.Errorf("internal error: cannot get snaps: %s", err)
	}
	// Migrate snapstate
	for name, raw := range snaps {
		var snapst patch62SnapState
		if err := json.Unmarshal([]byte(*raw), &snapst); err != nil {
			return err
		}
		if hasSnapdSnapID(snapst) && snapst.SnapType != string(snap.TypeSnapd) {
			snapst.SnapType = string(snap.TypeSnapd)
			data, err := json.Marshal(snapst)
			if err != nil {
				return err
			}
			newRaw := json.RawMessage(data)
			snaps[name] = &newRaw
			st.Set("snaps", snaps)
			// We can have at most one snapd snap
			break
		}
	}

	// migrate tasks' snap setup
...
```

Reply:
Thanks, looks good, applied! You're right about the "impossible state".